### PR TITLE
Fix: Installment Plan Schedule for purchases

### PR DESCRIPTION
**Pull request by VarunArora14**
Fixes: https://github.com/antiwork/gumroad/issues/1409

### Fix

Implemented a snapshotting mechanism that captures installment plan details at the time of the original purchase, ensuring payment consistency throughout the installment lifecycle.

#### Database Changes
- Added `installment_plan_id` to reference the original installment plan
- Added `installment_number_of_installments` to store the original number of installments
- Added `installment_payment_amounts_cents` to store the exact payment amounts as JSON
- Added proper indexes and foreign key constraints

#### Model Changes
**Purchase Model:**
- Added `belongs_to :installment_plan` association
- Added `before_create :capture_installment_plan_snapshot` callback
- Implemented `capture_installment_plan_snapshot` method to store plan details
- Enhanced `calculate_installment_payment_price_cents` to use snapshot data first, with fallback to current plan
- Updated `fetch_installment_plan` method with snapshot-first logic

#### Subscription Model:
- Updated `recurrence` method to use snapshot data when available
- Maintains backward compatibility with legacy subscriptions without installment plan

**Migration code added** for `purchase` table changes and creating index on `installment_plan_id` along with making it foreign key to `product_installment_plans` table (to it's `id` column for reference to product installment plan).

#### Existing installment calculation from `product_installment_plan.rb`

Existing code from `purchase.rb` invokes `fetch_installment.calculate_installment_payment_price_cents(total_price_cents)` which is defined as below -

```rb
  def calculate_installment_payment_price_cents(full_price_cents)
    base_price = full_price_cents / number_of_installments
    remainder = full_price_cents % number_of_installments

    Array.new(number_of_installments) do |i|
      i.zero? ? base_price + remainder : base_price
    end
  end
```

Current code changes **stores the installment plan to avoid recalculation of installment and directly fetches the array from the database** as per `capture_installment_plan_snapshot` method.

```
if installment_payment_amounts_cents.present?
        amounts = JSON.parse(installment_payment_amounts_cents)
        return amounts[nth_installment] || amounts.last
```

#### Testing

Added comprehensive test suite (`spec/models/installment_plan_snapshot_spec.rb`) covering:
- Snapshot capture functionality
- Payment amount calculation with remainders
- Fallback behavior for legacy purchases
- Edge cases and error handling

Tests Passing - 

image

AI Disclosure:

Copilot and Cursor used to understand functionality and help with specs creation

### CODE REVIEW

File: app/models/purchase.rb [hunk 1]
```diff
@@ -2642,6 +2644,26 @@ def transcode_product_videos
...
+      return unless is_installment_payment
+
+      # Determine which installment this is (0-indexed)
+      nth_installment = subscription&.purchases&.successful&.count || 0
+
+      # Use snapshot data if available (for purchases made after snapshotting was implemented)
+      if installment_payment_amounts_cents.present?
+        amounts = JSON.parse(installment_payment_amounts_cents)
+        return amounts[nth_installment] || amounts.last
+      end
```
Comment (VarunArora14) [inline app/models/purchase.rb:2657]: fetches `installment_payment_amounts_cents` created as per `capture_installment_plan_snapshot` method and fetches amounts array and then the nth installment

File: app/models/purchase.rb [hunk 2]
```diff
@@ -2642,6 +2644,26 @@ def transcode_product_videos
...
+
+      # Fallback to current plan for legacy purchases
+      current_plan = fetch_installment_plan
+      return unless current_plan
+
+      installment_amounts = current_plan.calculate_installment_payment_price_cents(total_price_cents)
+      installment_amounts[nth_installment] || installment_amounts.last
```
Comment (VarunArora14) [inline app/models/purchase.rb:2664]: fallback to use same `fetch_installment_plan` logic

File: app/models/purchase.rb [hunk 3]
```diff
@@ -3822,6 +3836,26 @@ def trigger_iffy_moderation
...
+      current_plan = fetch_installment_plan
+      return unless current_plan
+
+      # Calculate the total price for the installment calculation
+      # Use the price_cents before any fees are applied
+      total_price_cents = price_cents || 0
+
+      # Store the snapshot
+      self.installment_plan_id = current_plan.id
+      self.installment_number_of_installments = current_plan.number_of_installments
```
Comment (VarunArora14) [inline app/models/purchase.rb:3858]: creates plan for installment purchases (>1 installments), stores fields `installment_plan_id`, `installment_number_of_installments`, `installment_payment_amount_cents` (array which has calculated installments made to json)

This uses purchase's `number_of_installments` field to calculate installments and persists the `installment_payment_amounts_cents` array.

File: app/models/subscription.rb [hunk 4]
```diff
@@ -623,7 +623,13 @@ def formatted_end_time_of_subscription
 
   def recurrence
     if is_installment_plan
-      last_payment_option.installment_plan.recurrence
+      # Use snapshot data if available (for subscriptions created after snapshotting was implemented)
+      if original_purchase&.installment_plan_id.present?
+        original_purchase.installment_plan.recurrence
+      else
```
Comment (VarunArora14) [inline app/models/subscription.rb:629]: installment plan consideration for subscription as well along with fallback

File: db/schema.rb [hunk 5]
```diff
@@ -1739,13 +1739,17 @@
     t.bigint "price_id"
     t.string "recommended_by"
     t.datetime "deleted_at", precision: nil
+    t.bigint "installment_plan_id"
+    t.integer "installment_number_of_installments"
+    t.text "installment_payment_amounts_cents"
```
Comment (VarunArora14) [inline db/schema.rb:1744]: adds columns to "purchases" table and index, foreign key `installment_plan_id`

File: db/schema.rb [hunk 6]
```diff
@@ -2720,4 +2724,5 @@
 
   add_foreign_key "active_storage_attachments", "active_storage_blobs", column: "blob_id"
   add_foreign_key "active_storage_variant_records", "active_storage_blobs", column: "blob_id"
+  add_foreign_key "purchases", "product_installment_plans", column: "installment_plan_id"
```
Comment (VarunArora14) [inline db/schema.rb:2727]: foreign key `installment_plan_id` created to reference `ProductInstallmentPlan` that was active during time of purchase.

File: db/migrate/20251009171251_add_installment_plan_snapshot_to_purchases.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+class AddInstallmentPlanSnapshotToPurchases < ActiveRecord::Migration[7.1]
+  def change
+    add_column :purchases, :installment_plan_id, :bigint
+    add_column :purchases, :installment_number_of_installments, :integer
+    add_column :purchases, :installment_payment_amounts_cents, :text
+
+    add_index :purchases, :installment_plan_id
+    add_foreign_key :purchases, :product_installment_plans, column: :installment_plan_id
+  end
```
Comment (VarunArora14) [inline db/migrate/20251009171251_add_installment_plan_snapshot_to_purchases.rb:11]: Migration to add snapshot fields for installment plans to purchases.  Adds columns `instalment_plan_id`, `instalment_payment_amount_cents`, `installment_number_of_installments` to purchases table along with index on `installment_plan_id` and making it foreign key to table `product_installment_plans`

**Comment by VarunArora14** [comment]
RFR! 🙏🏻

**Comment by VarunArora14** [comment]
[@sla](https://github.com/slavingia) can you please provide feedback for this so I can improve this PR?
